pyanaconda/modules/localization/runtime.py

```diff
@@ -0,0 +1,178 @@
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+from pyanaconda.core.util import execWithRedirect
+from pyanaconda.core.constants import DEFAULT_KEYBOARD
+from pyanaconda.modules.common.errors.configuration import KeyboardConfigurationError
+from pyanaconda.modules.common.task import Task
+from pyanaconda.modules.localization.localed import LocaledWrapper
+from pyanaconda.anaconda_loggers import get_module_logger
+from pyanaconda.modules.localization.installation import write_x_configuration, \
+    write_vc_configuration
+
+log = get_module_logger(__name__)
+
+
+class ConvertMissingKeyboardConfigurationTask(Task):
+    """Task for getting missing keyboard settings by conversion."""
+
+    def __init__(self, keyboard, x_layouts, vc_keymap):
+        """Create a new task,
+
+        :param keyboard: generic system keyboard specification
+        :type keyboard: str
+        :param x_layouts: list of x layout specifications
+        :type x_layouts: list(str)
+        :param vc_keymap: virtual console keyboard mapping name
+        :type vc_keymap: str
+        """
+        super().__init__()
+        self._keyboard = keyboard
+        self._x_layouts = x_layouts
+        self._vc_keymap = vc_keymap
+
+    @property
+    def name(self):
+        return "Convert missing keyboard settings."
+
+    def run(self):
+        """Run conversion of missing keyboard settings.
+
+        :returns: tuple of X layouts and VC keyboard settings
+        :rtype: (list(str), str))
+        """
+        localed = LocaledWrapper()
+        vc_keymap = self._vc_keymap
+        x_layouts = self._x_layouts
+
+        new_vc_keymap = ""
+        if self._keyboard and not (vc_keymap or x_layouts):
+            # we were given just a value in the old format, use it as a vc_keymap
+            new_vc_keymap = self._keyboard
+        elif not vc_keymap and x_layouts:
+            new_vc_keymap = localed.convert_layouts(x_layouts)
+        elif not vc_keymap:
+            new_vc_keymap = DEFAULT_KEYBOARD
+
+        if new_vc_keymap:
+            vc_keymap = new_vc_keymap
+
+        if not x_layouts:
+            x_layouts = localed.convert_keymap(vc_keymap)
+
+        return x_layouts, vc_keymap
+
+
+class ApplyKeyboardTask(Task):
```

Comment:
I would try to move to this task the whole keyboard configuration from anaconda.py:

```

    configured = any((localization_proxy.Keyboard,
                      localization_proxy.VirtualConsoleKeymap,
                      localization_proxy.XLayouts))

    if opts.keymap and not configured:
        localization_proxy.SetKeyboard(opts.keymap)
        configured = True

    if configured:
        if conf.system.can_activate_keyboard:
            keyboard.activate_keyboard(localization_proxy)
        else:
            # at least make sure we have all the values
            keyboard.populate_missing_items(localization_proxy)
```

Reply:
After looking at the code I still think we want Apply and PopulateMissing API used by client. It is possible that the latter will be dropped? Or even handled by the client.